### PR TITLE
[AppGateServer] feat: Use round-robin endpoint selection strategy

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

The provided code implements a round-robin endpoint selection strategy for choosing a relayer endpoint from a list of valid endpoints. Here's a summary of how it works:

1. It filters the list of supplier endpoints to only include those that match the requested serviceId and rpcType. This filtered list is stored in validSupplierEndpoints.
2. If no valid endpoints are found, it returns an error (ErrAppGateNoRelayEndpoints).
3. It uses a mutex (app.endpointSelectionIndexMu) to protect the endpointSelectionIndex from concurrent modifications. This is necessary because multiple goroutines (representing multiple relay requests) could be trying to update the index at the same time.
4. It selects the next endpoint in the list by incrementing app.endpointSelectionIndex and taking the modulus of the number of valid endpoints. This ensures that the index always points to a valid endpoint in the list, and wraps around to the start of the list when it reaches the end.
4. It returns the selected endpoint.

This strategy ensures that all valid endpoints are used evenly, because each request selects the next endpoint in the list. However, it does not take into account the suitability of the endpoints, as noted in the TODO comment at the start of the function.

## Issue

For load testing purposes, the greedy endpoint selection strategy is no longer viable as it always use the first endpoint found while the tests need to distribute the relay request across the available `Appgate Server`s.

image

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR. **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
- [ ] **Documentation changes**: `make docusaurus_start`

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and referenced any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: pkg/appgateserver/endpoint_selector.go [hunk 1]
```diff
@@ -18,18 +18,35 @@ func (app *appGateServer) getRelayerUrl(
...
+
+	// Protect the endpointSelectionIndex update from concurrent relay requests.
+	app.endpointSelectionIndexMu.Lock()
+	defer app.endpointSelectionIndexMu.Unlock()
+
+	// Select the next endpoint in the list by rotating the index.
+	// This does not necessarily start from the first endpoint of a new session
+	// but will cycle through all valid endpoints of the same session if enough
+	// requests are made.
+	app.endpointSelectionIndex = (app.endpointSelectionIndex + 1) % len(validSupplierEndpoints)
```
Comment (Olshansk) [inline pkg/appgateserver/endpoint_selector.go:49]:
```suggestion
	// This is a naive strategy that is used to ensure all endpoints are leveraged throughout the
	// lifetime of the supplier. It is primarily used as a foundation for testing or development purposes
	// but a more enhanced strategy is expected to be adopted by prod gateways.
	app.endpointSelectionIndex = (app.endpointSelectionIndex + 1) % len(validSupplierEndpoints)
```

**Review by Olshansk — APPROVED** [review]
Awesome. Loved the explanation in the PR description too!

File: pkg/appgateserver/endpoint_selector.go (hunk 1 above)
Comment (h5law) [reply pkg/appgateserver/endpoint_selector.go:49]: Maybe change this to a `TODO_IMPROVE` otherwise LGTM

**Review by h5law — APPROVED** [review]
LGTM I think adding a comment somewhere as a TODO_IMPROVE for a future iteration will be good besides from that nice work!

**Pull request merged**